### PR TITLE
Provide configurable publish interface and support advertising "A" record on IPv4 or "AAAA" record on IPv6

### PR DESCRIPTION
**Pull request by LeZhang2016**

### CODE REVIEW

File: src/common/interface_address.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+private:
+    enum
+    {
+        kState_Ok       = 0,
+        kState_NotFound = -1,
+    };
+
+    char mIpv4Addr[20];
+    char mIpv6Addr[50];
```
Comment (bukepo) [inline src/common/interface_address.hpp]: magic number. use system defined `INET_ADDRSTRLEN` and `INET6_ADDRSTRLEN` instead?

File: src/common/interface_address.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+{
+    struct ifaddrs *ifaHead, *ifaNext;
+    int             index = 0, ret = kState_Ok;
+    bool            gotIpv4 = false;
+    bool            gotIpv6 = false;
+
+    if (getifaddrs(&ifaHead) == -1)
+    {
+        ret = kState_NotFound;
+        goto exit;
```
Comment (bukepo) [inline src/common/interface_address.cpp]: Use `VerifyOrExit()`?

File: src/common/interface_address.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    mInterfaceIndex = 0;
+    ifaNext = ifaHead;
+    while (ifaNext)
+    {
+        if (ifaNext->ifa_addr && ifaNext->ifa_addr->sa_family == AF_PACKET)
+        {
+            index++;
+            if (strcmp(ifaNext->ifa_name, aInterfaceName) == 0)
+            {
+                mInterfaceIndex = index;
```
Comment (bukepo) [inline src/common/interface_address.cpp]: If these code are aimed at finding index of a interface, can it be done by POSIX API `if_nametoindex()`?

File: src/common/interface_address.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            if ((ifaNext->ifa_addr) && (ifaNext->ifa_addr->sa_family == AF_INET) && !gotIpv4)
+            {
+                struct sockaddr_in *in = (struct sockaddr_in *) ifaNext->ifa_addr;
+                inet_ntop(AF_INET, &in->sin_addr, mIpv4Addr, sizeof(mIpv4Addr));
+                gotIpv4 = true;
+            }
+            else if ((ifaNext->ifa_addr) && (ifaNext->ifa_addr->sa_family == AF_INET6) && !gotIpv6)
+            {
+                struct sockaddr_in6 *in6 = (struct sockaddr_in6 *) ifaNext->ifa_addr;
```
Comment (bukepo) [inline src/common/interface_address.cpp]: `static_cast<>` instead?

File: src/utils/hex.cpp [hunk 5]
```diff
@@ -127,6 +127,38 @@ int Long2Hex(const uint64_t aLong, char *aHex)
...
+        }
+        else
+        {
+            index++;
+        }
+        cur++;
+    }
+}
+
+void  Ipv6AddressToBytes(const char *aStr, uint8_t *aAddress, uint8_t aLen)
```
Comment (bukepo) [inline src/utils/hex.cpp]: can this be replaced by `inet_pton(AF_INET6, ...)`?

File: src/utils/hex.cpp [hunk 6]
```diff
@@ -127,6 +127,38 @@ int Long2Hex(const uint64_t aLong, char *aHex)
     return strlen(aHex);
 }
 
+
+void Ipv4AddressToBytes(const char *aStr, uint8_t *aAddress, uint8_t aLen)
```
Comment (bukepo) [inline src/utils/hex.cpp]: can this be replaced by `inet_pton(AF_INET, ...)`?

**Comment by jwhui** [comment]
@LeZhang2016, what is the status of this PR?

**Comment by jwhui** [comment]
This PR has gone stale. Closing.

**Pull request closed without merging**